Adds documentation about client-side muting. (#504)
medic/cht-docs: content/en/apps/reference/app-settings/transitions.md

## Configuration

By default all transitions are disabled. They can be enabled by configuring the `transitions` property to have a key with the transitions name and a `truthy` value. 

> As of version 3.12.0, some transitions can partially run on the client, for offline users. To op out from client-side transitions, add a `"client: false"` property to the transition configuration.

Reviewer: ```suggestion
As of version 3.12.0 some transitions will partially run on the client for offline users. To opt out from client-side transitions, add a `"client: false"` property to the transition configuration.
```

[1] Contacts that are already in the correct state are skipped. This applies to updates to the contact itself, updates to the Sentinel `muting_history` and to the connected registrations (registrations of a contact that is already in the correct state will not be updated).  
[2] The date represents the moment Sentinel has processed the muting action  
[3] target contact and descendants' registrations  
[4] `scheduled_tasks` being in either `scheduled` or `pending` state  
[5] Because the muted state is inherited, unmuting cascades upwards to the highest level muted ancestor. If none of the ancestors is muted, unmuting cascades downwards only.  
[6] `scheduled_tasks` which are due today or in the future. All `scheduled_tasks` with a due date in the past will remain unchanged.  
[8] The date represents the device's date when the report is processed.  
[9] Replaying is required due to how PouchDB <-> CouchDB synchronization does not respect the order in which the documents have been created, to ensure that contacts end up in the correct muted state.    

> [10] The updated contacts are limited to the contacts available on the device.  

Reviewer 2: I hadn't noticed our use of footnotes before, and find it quite hard to find/use given that they aren't cross linked (eg it isn't clear where to go to see the footnotes, and 7 is separate). It may be outside of the scope of this PR to fix, but just highlighting as we are adding to them here.

Author: Yea, I think with plain markdown files it wasn't possible to superscript and also link, maybe we have some options now.

Reviewer 2: I'll create a separate issue for the footnotes.

Reviewer 2: Closing the loop: https://github.com/medic/cht-docs/issues/509

| date | Date in ISO Format |
| report_id | An `_id` reference to the report that triggered the action |

> ##### [11] Client-side Muting history as of 3.12.0

Reviewer 3: worth noting in contacts schema page?
https://docs.communityhealthtoolkit.org/core/overview/db-schema/#contacts-persons-and-places

Author: Unsure. This is an implementation detail, I wouldn't want it: a. edited or b. used for any other purpose by data-analysis and such.

Reviewer 3: it is in the contact doc, right? it is going to get used :( 
if that's not a good idea - maybe that's what you should say in the schema documentation?

Author: It is in the contact doc, correct. We need this metadata to insure some sort of consistency with out of order syncing and muting. The alternative would have been to have another doc type to sync to just contain this metadata.

Author: Server-side Muting history exists in infodocs, and can be used for dashboards.

Reviewer 3: > The alternative would have been to have another doc type to sync to just contain this metadata

Sounds not that nice.

I'm not really saying you shouldn't put the data here. But I'm saying that this data is very public and discoverable, so it is likely it will be used.

Author: I can't guarantee that this data structure won't change over time. It being an implementation artifact, I don't want to block future development from changing this structure to accommodate new functionalities or bug fixes because data analysis is already using it. 
How do you think it's best communicated that this data is not guaranteed to have this format or exist at all?

Reviewer 3: most people writing data analytic queries are doing it from xform data. with that data, all the internal calculations/intermediate steps/UI strings are even exposed publicly. it is very common for a query to depend on something fragile which either a) breaks the query, or b) makes it more difficult to edit the form later.

we would benefit from a convention of "private vs public data", but i believe we'd be creating it here. today, everything is fair game cc @derickl

Each time the client changes the `muted` state of a contact, an entry is added to a `muting_history` property on the contact's doc. The `last_update` entry is also changed to `client`.
The `muting_history` property contains the following information:

| Property | Values | Description | 
| --- | ---- | ---- |
| last_update | `server` or `client` | Updated every time a service updates the contact, with the corresponding value | 
| server | Object |  |
| server.muted | `true` or `false` | Last known server muting state |
| server.date | Date in ISO format | Last known server muting/unmuting date | 

> | client | Array | Client-side muting/unmuting events list. <br> New events are pushed at the end of this list and it should never be re-ordered. <br>The list represents the "chronological" order in which the reports that triggered muting were created. | 

Reviewer 2: Just a note to update this based on the result of this conversation.

Author: Editing this would require quite a large refactor, I was just planning to edit the config.

Author: This part: https://github.com/medic/cht-docs/pull/504/files#diff-03618adbb0b596d7e314daf185764395f0b911b479b72ac9954722007b776c93R25

Reviewer 2: Ah, I see. Would other potential future client-side transitions go here as well? Does having client here and client_side on the config make it more confusing? I'll let you make a call on that with app developers.

Author: It depends, right now this is an implementation detail for muting specifically, which is in itself a complex piece of work. It's used exclusively to have consistent muting while having out of order syncs. 
App developers or anyone really should be using these.

Author: But I'll do the change anyway :P
